File: src/main/java/com/ghouse/sana/proxy/GptProxy.java
package com.ghouse.sana.proxy;

import com.ghouse.sana.dto.GptRQ;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.*;
import org.springframework.stereotype.Component;
import org.springframework.web.client.RestTemplate;

@Component
public class GptProxy {
    private final RestTemplate restTemplate = new RestTemplate();
    private static Logger log = LoggerFactory.getLogger(GptProxy.class);

    @Value("${openai.api.key}")
    private String apiKey;

    public String sendRequest(GptRQ gptRQ) {

        String endpoint = "https://api.openai.com/v1/chat/completions";

        HttpHeaders headers = new HttpHeaders();
        headers.setContentType(MediaType.APPLICATION_JSON);
        headers.setBearerAuth(apiKey);

        try {
            HttpEntity<GptRQ> requestEntity = new HttpEntity<>(gptRQ, headers);
            ResponseEntity<String> responseEntity = restTemplate.exchange(endpoint, HttpMethod.POST, requestEntity, String.class);

            System.out.println("API Response Code: " + responseEntity.getStatusCode());
            System.out.println("API Response Body: " + responseEntity.getBody());
            return responseEntity.getBody();

        } catch (Exception exp) {
            log.error("Exception when calling api: {} ", exp.getMessage());
            throw exp;
        }
    }
}

File: proxy/gptProxy.js
import dotenv from 'dotenv';
import axios from 'axios';
dotenv.config();
export default class GptProxy {
    constructor() {
        this.apiKey = process.env.OPENAI_API_SECRET;
        if (!this.apiKey) {
            throw new Error('OPENAI_API_KEY environment variable is not set.');
        }
    }

    async sendRequest(gptRQ) {
        const endpoint = "https://api.openai.com/v1/chat/completions";

        const headers = {
            'Content-Type': 'application/json',
            'Authorization': `Bearer ${this.apiKey}`
        };

        try {
            const response = await axios.post(endpoint, gptRQ, { headers });
            
            console.log("API Response Code: " + response.status);
            console.log("API Response Body: " + response.data);
            return response.data;
        } catch (error) {
            if (error.response) {
                console.error("API Error Response:");
                console.error("Status:", error.response.status);
                console.error("Data:", error.response.data);
            } else if (error.request) {
                console.error("No API Response:", error.request);
            } else {
                console.error("Error:", error.message);
            }
            throw error;
        }
        
    }
}
